cpp/src/reductions/scan/scan_inclusive.cu

```diff
@@ -198,22 +196,210 @@ struct scan_dispatcher {
+        bool const rhs_is_null{d_order_by.is_null_nocheck(row_index - 1)};
+        if (lhs_is_null and rhs_is_null) {
+          return 0;
+        } else if (lhs_is_null != rhs_is_null) {
+          return row_index + 1;
+        }
+      }
+
+      return thrust::all_of(
+               thrust::seq,
+               thrust::make_counting_iterator<size_type>(0),
+               thrust::make_counting_iterator<size_type>(0) + d_order_by.num_child_columns(),
+               [row_index, d_order_by] __device__(size_type child_index) {
+                 column_device_view const& col = d_order_by.child(child_index);
+                 element_equality_comparator<nested_nulls> element_comparator{col, col, true};
+                 return type_dispatcher(col.type(), element_comparator, row_index, row_index - 1);
+               })
+               ? 0
+               : row_index + 1;
+    });
+}
+
+template <bool has_nulls>
+void generate_rank_comparisons(column_view const& order_by,
+                               mutable_column_view out,
+                               rmm::cuda_stream_view stream)
+{
+  auto const d_order_by = column_device_view::create(order_by, stream);
+  element_equality_comparator<has_nulls> element_comparator(*d_order_by, *d_order_by, true);
+  thrust::tabulate(rmm::exec_policy(stream),
+                   out.begin<size_type>(),
+                   out.end<size_type>(),
+                   [type = d_order_by->type(), element_comparator] __device__(size_type row_index) {
+                     return (row_index == 0 ||
+                             !type_dispatcher(type, element_comparator, row_index, row_index - 1))
+                              ? row_index + 1
+                              : 0;
+                   });
+}
+
+template <bool nested_nulls>
+void generate_dense_rank_struct_comparisons(column_view const& order_by,
+                                            mutable_column_view out,
+                                            rmm::cuda_stream_view stream)
+{
+  auto const d_order_by = column_device_view::create(order_by, stream);
+  thrust::tabulate(
+    rmm::exec_policy(stream),
+    out.begin<size_type>(),
+    out.end<size_type>(),
+    [d_order_by = *d_order_by, has_nulls = order_by.has_nulls()] __device__(size_type row_index) {
+      if (row_index == 0) {
+        return true;
+      } else if (has_nulls) {
+        bool const lhs_is_null{d_order_by.is_null_nocheck(row_index)};
+        bool const rhs_is_null{d_order_by.is_null_nocheck(row_index - 1)};
+        if (lhs_is_null and rhs_is_null) {
+          return false;
+        } else if (lhs_is_null != rhs_is_null) {
+          return true;
+        }
+      }
+
+      return !thrust::all_of(
+        thrust::seq,
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(0) + d_order_by.num_child_columns(),
+        [row_index, d_order_by] __device__(size_type child_index) {
```

Comment:
Isn't there an iterator over children of a column?

Reply:
Right, we should use children iterators here too.

Reply:
There is for `column_view`'s, but not `column_device_view`'s. Since we're nested inside a `thrust::tabulate` and inside of device code, I didn't see a clean way to access those iterators (`child_begin()` and `child_end()`).

Reply:
Sounds like we just need to add `child_begin/end` to `column_device_view`.

Reply:
To clarify, it sounds like Jake is asking @rwlee to add these in this PR. I'm clarifying because we are extremely close to code freeze.

Reply:
I have a proposed alternative approach that removes the need for this iterator and a few other issues brought up by @ttnghia and @jrhemstad. It would allow the addition of child_begin/end to be pushed to a later release and added in as a separate feature.